### ISSUE TITLE
Enhance collision detection performance with quadtree-indexed Canvas Edges

### ISSUE THREAD

**Issue by Codas** [foundryvtt/foundryvtt#11798]
### User Experience

Hey foundry team :)

While investigating some performance issues especially with the PF2e aura drawing code (more especially in combination with the levels module), it came to my attention that `PointSourcePolygon.testCollision` always checks every canvas edge for collision.
The code in question is actually in the instance method `identifyEdges`, which is called while executing `testCollision`:
```js
_identifyEdges() {
  /* .. */
  for (const edge of canvas.edges.values()) {
     /* .. */
  }
}
```
Instead of iterating over all the edges, a quadtree indexing of the edges could be used to drastically reduce the amount of edges that need to be checked.

In my testing, using a quadtree to get only the edges needed for collision checking or clockwiseSweepPolygon creation (+additional canvas bounds edges) drastically reduces the time needed for the collision checks.

In a scene with 1099 walls, a 90ft aura takes about 195ms to calculate with the current implementaiton, whereas an implementation of `_identifyEdges` that uses the rays bounds to only query releveant edges reduces this to 27.5ms on my system.

I have hacked together a [proof of concept](https://github.com/Codas/foundryvtt-performance-hacks/blob/main/src/hacks/faster-edge-testing.ts) in my prime Performance module that demonstrates this approach. I belive the positive effects outweigh the slightly increased memory footprint for the quadtree and minmal overhead when querying all edges for basic vision quadtree calculations. The overhead when querying all edges in the scene for for vision calculation is about 15% (when animating token movement for example), or an increase from 1.45ms to 1.7ms on my system in this test scene.

A more sophisticated implementation could check for the bounds and only use the quadtree indexing when the requested bounds do not completely encase the outer canvas bounds to reduce this overhead.

**Comment by aaclayton** [comment]
Worth consideration, I had initially not implemented a quadtree here because in my own testing the performance difference was negligible even for large scenes except for very small movements. As soon as you have a large proposed movement I had found that the performance cost between using a quadtree to identify edges vs. iterating over all of them was relatively minor. I think it's worth revisiting though. I don't know if we'll have the bandwidth in v13 though with everything else on our plates right now.

**Comment by Codas** [comment]
Thank you for taking it into consideration, that's all I can ask for :)

Core foundry would probably not benefit from it as much as some modules or systems could that heavily rely on the `testCollision` function as opposed to creating a full vision polygon, that is true.

**Comment by aaclayton** [comment]
Depending on your use case, you _may_ find it more efficient to create a single `ClockwiseSweepPolygon` (or alternative polygon backend) with `type: "move"` and then use the resulting polygon for further tests than doing a lot of individual `testCollision` calls.

**Comment by Codas** [comment]
Exactly! It's more efficient once about 50 or so tests are being made against the same origin.
Thats also the techinique I used to improve the pf2e aura rendering performance.
But for other cases where a few checks are being made per origin but for many different sources a more performant `testCollision` would still be nice to have.

That's exactly it though. It would be a nice to have but it's also not super important. So as you said: You're considering it, and that's all I wanted.
